central/image/datastore/store/v2/postgres/store.go

```diff
@@ -0,0 +1,1030 @@
+		if err != nil {
+			return err
+		}
+	}
+
+	// If the scan is not new, we do not need to bother writing the components and CVEs as the latest already
+	// exist.
+	if !scanUpdated {
+		common2.SensorEventsDeduperCounter.With(prometheus.Labels{"status": "deduped"}).Inc()
+		return nil
+	}
+	common2.SensorEventsDeduperCounter.With(prometheus.Labels{"status": "passed"}).Inc()
+
+	err = s.copyFromImageComponentsV2(ctx, tx, parts.image.GetId(), parts.componentsV2...)
+	if err != nil {
+		return err
+	}
+
+	return copyFromImageComponentV2Cves(ctx, tx, iTime, cveTimeMap, parts.cveV2...)
+}
+
+func getPartsAsSlice(parts common.ImageParts) *imagePartsAsSlice {
+	componentsV2 := make([]*storage.ImageComponentV2, 0, len(parts.Children))
+	vulns := make([]*storage.ImageCVEV2, 0)
+	for _, child := range parts.Children {
+		componentsV2 = append(componentsV2, child.ComponentV2)
+		for _, gChild := range child.Children {
+			vulns = append(vulns, gChild.CVEV2)
+		}
+	}
+	return &imagePartsAsSlice{
+		image:        parts.Image,
+		componentsV2: componentsV2,
+		cveV2:        vulns,
+	}
+}
+
+func insertIntoImagesLayers(ctx context.Context, tx *postgres.Tx, obj *storage.ImageLayer, imageID string, idx int) error {
+	values := []interface{}{
+		// parent primary keys start
+		imageID,
+		idx,
+		obj.GetInstruction(),
+		obj.GetValue(),
+	}
+
+	finalStr := "INSERT INTO images_Layers (images_Id, idx, Instruction, Value) VALUES($1, $2, $3, $4) ON CONFLICT(images_Id, idx) DO UPDATE SET images_Id = EXCLUDED.images_Id, idx = EXCLUDED.idx, Instruction = EXCLUDED.Instruction, Value = EXCLUDED.Value"
+	_, err := tx.Exec(ctx, finalStr, values...)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *storeImpl) copyFromImageComponentsV2(ctx context.Context, tx *postgres.Tx, imageID string, objs ...*storage.ImageComponentV2) error {
+	// Each scan is complete replacement.  So first thing we do is remove the old components and CVEs for an image.
+	if err := s.deleteImageComponents(ctx, tx, imageID); err != nil {
+		return err
+	}
+	batchSize := pgSearch.MaxBatchSize
+	if len(objs) < batchSize {
+		batchSize = len(objs)
+	}
+	inputRows := make([][]interface{}, 0, batchSize)
+
+	copyCols := []string{
+		"id",
+		"name",
+		"version",
+		"priority",
+		"source",
+		"riskscore",
+		"topcvss",
+		"operatingsystem",
+		"imageid",
+		"location",
+		"serialized",
+	}
+
+	for idx, obj := range objs {
+		serialized, marshalErr := obj.MarshalVT()
+		if marshalErr != nil {
+			return marshalErr
+		}
+
+		inputRows = append(inputRows, []interface{}{
+			obj.GetId(),
+			obj.GetName(),
+			obj.GetVersion(),
+			obj.GetPriority(),
+			obj.GetSource(),
+			obj.GetRiskScore(),
+			obj.GetTopCvss(),
+			obj.GetOperatingSystem(),
+			obj.GetImageId(),
+			obj.GetLocation(),
+			serialized,
+		})
+
+		// if we hit our batch size we need to push the data
+		if (idx+1)%batchSize == 0 || idx == len(objs)-1 {
```

Comment:
nit: I won't lie, I'm no expert on micro-optimizations, but I have been led to believe `%` can be costly, especially compared to just an equality check. So, I'm wondering if instead of `idx` we just track a separate variable that counts up to `batchSize` and we reset it once we write. Not sure if we do enough operations here to make this noticeable or not, though, so you may be able to ignore this

Reply:
We do this in many places.  I'm going to defer for now, but I may write a ticket (or I may not) and take a look later.  I am intrigued by this.